### PR TITLE
ffmpeg fixes

### LINKED ISSUE #44: SWF may be not user friendly extension for timelapse which is just MPEG2 file
Is it possible to use SWF codec while file extension is .MPEG ?
Most browsers detect .swf extension and try to load flash player which will mostly fail. If the file is just MPEG2 would make sense to use video/mpeg mime type. 

So I would suggest:
1. make SWF default to protect against overwrite
2. if codec is SWF, append extension ".mpeg"
3. if other codec, append ".avi" or ".mpg" accordingly

### LINKED ISSUE #56: Variable FPS bug
(cross post from mail list)

Hi,

I have seen this bug for quite long time, but now finally created
proof of concept.

The issue is how motion deals with variable frame rate. Depending on
the amount of light a web cam changes FPS: more light - more frames
per second.

When real frame rate increases - motion generates "slow" movie. For
example, motion started movie as 7 fps, but then camera started
sending 30 fps - you will have "slowmotion" effect.

When real frame rate decreases - motion starts filling "pauses" into
movie: copying last second's frame N-times.

Please see examples: (download original MP4, don't use flash player)

"Slowmotion" - https://yadi.sk/i/hnsMAPhTeKpH9
"Lags" - https://yadi.sk/i/4cnHPQAyeKpKj

Config: http://pastebin.com/FUViNzQt
Uses extpipe x264 (compiled with gpac)
Compiled latest "master" branch of Mr-Dave - https://github.com/Mr-Dave/motion

To reproduce "slowmotion", close web camera lens using your finger -
it will drop to the lowest framerate. Start motion, wait for 10
seconds and start generating movements - motion will begin writing a
movie at 7 fps. Then use flashlight to bring more light into web cam
lens - camera starts giving maximum frame rate.

To reproduce "lags" do vice versa. Turn on flashlight, camera will
switch to highest frame rate. Start motion and generate movements -
motion starts movie at 30 fps. Now turn flashlight off - camera drops
frame rate to lover (15 in my case).

Any ideas how we could fix that?

Thanks

### LINKED ISSUE #113: Wrong range for ffmpeg_variable_bitrate option
Playing with ffmpeg_variable_bitrate I noticed it had no effect whatsoever. Digging in deeper I discovered that the actual range for the quality used by ffmpeg/libav/whatever is 1 to 32767 (see [AVFrame::quality](https://www.ffmpeg.org/doxygen/3.0/structAVFrame.html#a791c071afd3e95fcae14cb781829754d) and [FF_LAMBDA_MAX](https://www.ffmpeg.org/doxygen/3.0/group__lavu__enc.html#gabc22a7ffbd4a73e81dd69e8c45e6cf8c)). Whereas the motion docs say valid values are 0 and 2 to 31.

Now I have tested higher values and indeed, at say 31000, the movie quality is pretty bad (expected). I believe somehow, somewhere there's a missing factor of 1000.

Should the docs be updated or should motion be patched to multiply the values by 1000 (or something)?

I'm not sure if this is the best place to open/discuss this issue. Anyways, I believe however that @tosiara, @sackmotion and @Mr-Dave may have a good input on this matter.

### LINKED ISSUE #131: Truncated images in netcam
If image size is over 1M then it gets truncated within netcam.c due to the inline coding <code> remaining = 999999</code>.   Originally reported here http://www.lavrsen.dk/foswiki/bin/view/Motion/BugReport2015x10x24x203225

### LINKED ISSUE #136: Deprecated ffmpeg functions
Following warnings appear during compilation:
```
gcc -g -O2 -D_REENTRANT -DMOTION_V4L2 -DMOTION_V4L2_OLD -DTYPE_32BIT="int" -DHAVE_BSWAP   -march=native -mtune=native -Wall -DVERSION=\"Unofficial-Git-2caced3\" -Dsysconfdir=\"/usr/local/etc\" -I/home/ubuntu/git/ffmpeg/out/include     -c -o netcam_rtsp.o netcam_rtsp.c
netcam_rtsp.c: In function ‘decode_packet’:
netcam_rtsp.c:148:5: warning: ‘avpicture_get_size’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4914) [-Wdeprecated-declarations]
netcam_rtsp.c:152:5: warning: ‘avpicture_layout’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4906) [-Wdeprecated-declarations]
netcam_rtsp.c: In function ‘netcam_read_rtsp_image’:
netcam_rtsp.c:316:13: warning: ‘av_free_packet’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4056) [-Wdeprecated-declarations]
netcam_rtsp.c:325:9: warning: ‘av_free_packet’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4056) [-Wdeprecated-declarations]
netcam_rtsp.c:333:5: warning: ‘av_free_packet’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4056) [-Wdeprecated-declarations]
netcam_rtsp.c: In function ‘netcam_rtsp_open_sws’:
netcam_rtsp.c:578:5: warning: ‘avpicture_get_size’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4914) [-Wdeprecated-declarations]
netcam_rtsp.c: In function ‘netcam_rtsp_resize’:
netcam_rtsp.c:614:5: warning: ‘avpicture_fill’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4899) [-Wdeprecated-declarations]
netcam_rtsp.c:632:5: warning: ‘avpicture_fill’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4899) [-Wdeprecated-declarations]
netcam_rtsp.c:664:5: warning: ‘avpicture_layout’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4906) [-Wdeprecated-declarations]
gcc -g -O2 -D_REENTRANT -DMOTION_V4L2 -DMOTION_V4L2_OLD -DTYPE_32BIT="int" -DHAVE_BSWAP   -march=native -mtune=native -Wall -DVERSION=\"Unofficial-Git-2caced3\" -Dsysconfdir=\"/usr/local/etc\" -I/home/ubuntu/git/ffmpeg/out/include     -c -o ffmpeg.o ffmpeg.c
ffmpeg.c: In function ‘ffmpeg_put_frame’:
ffmpeg.c:571:13: warning: ‘av_free_packet’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4056) [-Wdeprecated-declarations]
ffmpeg.c:588:5: warning: ‘av_free_packet’ is deprecated (declared at /home/ubuntu/git/ffmpeg/out/include/libavcodec/avcodec.h:4056) [-Wdeprecated-declarations]
ffmpeg.c: In function ‘ffmpeg_open’:
ffmpeg.c:397:30: warning: ignoring return value of ‘avformat_write_header’, declared with attribute warn_unused_result [-Wunused-result]
```

### PR DESCRIPTION
**Pull request by Mr-DaveDev**
These commits address multiple issues with the ffmpeg processing.  

1. The variable bit rate parameter will control quality for all the codec/container options and will close #113.  
2. The presentation time stamps (pts) were not being set right for all the different containers which closes #56 and #132.  
3. Fixed the truncated images which closes #131.  
4. The ogg container was not able to be successfully implemented and was removed as a available option.  Specification of this container will result in Motion falling back to the default.   
5. Revisions have been included to handle the ffmpeg 3.0 changes and remove the depreciation warnings.  Closes #136 
6. The swf container for the timelapse needed to be changed to mpg.  Using a codec of mpeg2 with a swf container is not permitted.  Other container options for timelapse not permitted so this container change closes #44

### CODE REVIEW

File: event.c [hunk 1]
```diff
@@ -627,9 +629,54 @@ static void event_ffmpeg_newfile(struct context *cnt,
      *  motion movies get the same name as normal movies plus an appended 'm'
      *  PATH_MAX - 4 to allow for .mpg to be appended without overflow
      */
-    snprintf(cnt->motionfilename, PATH_MAX - 4, "%s/%sm", cnt->conf.filepath, stamp);
-    snprintf(cnt->newfilename, PATH_MAX - 4, "%s/%s", cnt->conf.filepath, stamp);
-
+     /* Set up a testing / evaluation codec which will use a different
+      * codec for the events.
+     */
```
Comment (jogu) [inline event.c]: I'd possibly expand the comment a little to explain that setting the codec to "test" means that it cycles through all the codecs, switching at the end of each event.

File: event.c [hunk 2]
```diff
@@ -627,9 +629,54 @@ static void event_ffmpeg_newfile(struct context *cnt,
...
+    snprintf(codec, sizeof(codec), "%s", cnt->conf.ffmpeg_video_codec);
+    if (strcmp(codec, "test") == 0) {
+        MOTION_LOG(NTC, TYPE_ENCODER, NO_ERRNO, "%s Running test of the various output formats.");
+        codenbr = cnt->event_nr;
+        while (codenbr > 10){
+          codenbr -= 10;
+        }
+        switch (codenbr) {
+        case 1:
+            snprintf(codec, sizeof(codec), "%s", "mpeg4");
```
Comment (jogu) [inline event.c]: This code seems quite verbose; unless I've missed something I think "codec" could be declared as a const char *. Then this snprintf line can become simply codec = "mpeg4"; (and the line before the switch can be codec = cnt->conf.ffmpeg_video_codec)

File: event.c [hunk 3]
```diff
@@ -627,9 +629,54 @@ static void event_ffmpeg_newfile(struct context *cnt,
...
-
+     /* Set up a testing / evaluation codec which will use a different
+      * codec for the events.
+     */
+    snprintf(codec, sizeof(codec), "%s", cnt->conf.ffmpeg_video_codec);
+    if (strcmp(codec, "test") == 0) {
+        MOTION_LOG(NTC, TYPE_ENCODER, NO_ERRNO, "%s Running test of the various output formats.");
+        codenbr = cnt->event_nr;
+        while (codenbr > 10){
+          codenbr -= 10;
```
Comment (jogu) [inline event.c]: I think the while loop could be replaced with codenbr = cnt->event_nr % 10; ?

File: event.c [hunk 4]
```diff
@@ -700,7 +700,7 @@ static void event_ffmpeg_timelapse(struct context *cnt,
     if (!cnt->ffmpeg_timelapse) {
```
Comment (jogu) [inline event.c]: I'd possibly add a few notes to the commit message to this one about why swf is being named etc.

**Comment by jogu** [comment]
I've made a few comments on a couple of bits of the diff; generally this seems fine to me.

**Comment by Mr-DaveDev** [comment]
Yes, it is very verbose in code and lacking in comment :) I'll see about cleaning it up.  I wasn't initially planning on creating it or keeping it in but the testing was really painful doing it manually.   All these unusual / antiquated containers really needs to be re-considered.

> Maybe we should keep swf for compatibility with older configs and show a warning that swf is deprecated and will be removed soon, please use mpg instead?

I'll revise this to change the swf and ogg to new defaults with a WRN level.

**Comment by Mr-DaveDev** [comment]
Test and merge if you agree.

**Comment by tosiara** [comment]
I have compiled and running this tree. Looks good to merge

**Pull request merged**